sum.go

```diff
@@ -1,241 +1,49 @@
 package multihash
 
 import (
-	"crypto/md5"
-	"crypto/sha1"
-	"crypto/sha512"
 	"errors"
 	"fmt"
-
-	blake2b "github.com/minio/blake2b-simd"
-	sha256 "github.com/minio/sha256-simd"
-	murmur3 "github.com/spaolacci/murmur3"
-	blake2s "golang.org/x/crypto/blake2s"
-	sha3 "golang.org/x/crypto/sha3"
 )
 
 // ErrSumNotSupported is returned when the Sum function code is not implemented
 var ErrSumNotSupported = errors.New("Function not implemented. Complain to lib maintainer.")
 
 var ErrLenTooLarge = errors.New("requested length was too large for digest")
 
-// HashFunc is a hash function that hashes data into digest.
-//
-// The length is the size the digest will be truncated to. While the hash
-// function isn't responsible for truncating the digest, it may want to error if
-// the length is invalid for the hash function (e.g., truncation would make the
-// hash useless).
-type HashFunc func(data []byte, length int) (digest []byte, err error)
-
-// funcTable maps multicodec values to hash functions.
-var funcTable = make(map[uint64]HashFunc)
-
 // Sum obtains the cryptographic sum of a given buffer. The length parameter
 // indicates the length of the resulting digest and passing a negative value
 // use default length values for the selected hash function.
 func Sum(data []byte, code uint64, length int) (Multihash, error) {
-	if !ValidCode(code) {
-		return nil, fmt.Errorf("invalid multihash code %d", code)
+	// Get the algorithm.
+	hasher, err := GetHasher(code)
+	if err != nil {
+		return nil, err
 	}
 
+	// Feed data in.
+	hasher.Write(data)
+
+	// Compute hash.
+	//  Use a fixed size array here: should keep things on the stack.
```

Comment:
It won't due to interface indirections (go's escape analysis won't be able to deal with this and will put everything on the heap) and if we don't size this appropriately, CID allocations will 2x.

Let's allocate what we need (use the size hints).

Reply:
Yeah, I was smitten with wishful thinking here -- I _thought_ had heard something about the compiler getting clever enough to sometimes see through interfaces and avoid the escapes.  I was probably thinking of [this](https://twitter.com/mdempsky/status/1321894360116420608?s=09)/[this](http://golang.org/issue/33160), though, which is of limited scope and indeed doesn't apply here.

I'll drop this to just using 'nil' again so the hash function auto-sizes it.  Same effect.